### DISCUSSION TITLE
Partial object write for JSON string.

### DISCUSSION THREAD

**Discussion by wenzhengresideo** [stephenberry/glaze#204] [Q&A] [answered]
Hi is there a way to write out a partial json string? I see you can do it for binary output but would be very useful for as string as well.

    static constexpr auto partial = glz::json_ptrs("/a","/c");
    struct mystruct{int a, int b, int c};
    mystruct s{};
    std::string out;
    
    glz::write_json<partial>(s, out);

**Comment by mwalcott3** [comment]
If I remember correctly it was not added because unlike the binary spec there was not a good way of handling a partial where and element in an array was requested.

Take the structure:
```json 
{
  "foo": [0,1,2,3,4,5,6,7,8,9],
  "bar": true,
  "biz": {
    "x": 1.2,
    "y": 3.7,
    "z": 1009,
    "w": 3.14
  }
}
```

The following partial would be straightforward and obvious.
```c++
static constexpr auto partial = glz::json_ptrs("/bar","/biz/y");
```
```json 
{
  "bar": true,
  "biz": {
    "y": 3.7
  }
}
```

But lets say I include an array index in the json pointer
```c++
static constexpr auto partial = glz::json_ptrs("/foo/5","/biz/y");
```
What does the partial message look like?
Do we write out the entire array? 
Do we write all the values up to the one requested?
Do we write all the values but write null for items not requested in the array? or empty objects for an array of objects?
Do we flatten the structure and write the json pointer path to the item as the key?
Do we ban array access in partials?

Writing out the whole array results in allot of extra stuff
```json 
{
  "foo":  [0,1,2,3,4,5,6,7,8,9],
  "biz": {
    "y": 3.7
  }
}
```
The following solves the problem but is not really standard json for handling this
```json 
{
  "/biz/y":  3.7,
  "/foo/5": 4
}
```

As a side note:
glz::skip (Always ignore) and glz::hide (Read it in but don't write it out) can be used for some of the use cases for partials but do not allow having different outputs for the same objects.

I agree that being able to customize messages would be nice but am unsure of the best way to enable this. I'm open to api suggestions.

**Comment by wenzhengresideo** [comment]
Hi I guess one way would be to state that traversal of the json structure is only valid with string keys. I don't view an index of an array as an explicit key of json. So in the case of "/foo/5" it should be invalid. 
"/foo/5" means
{
"foo":
{"5": ...}
} 
which doesn't exist

**Reply by mwalcott3** [reply]
Seems reasonable. @stephenberry thoughts?

**Reply by stephenberry** [reply]
Yeah, I think this makes sense.

**Comment by stephenberry** [comment, marked as answer]
Partial JSON write support has been added in [v2.2.2](https://github.com/stephenberry/glaze/releases/tag/v2.2.2) (#815)

```c++
struct animals_t
{
   std::string lion = "Lion";
   std::string tiger = "Tiger";
   std::string panda = "Panda";

   struct glaze
   {
      using T = animals_t;
      static constexpr auto value = glz::object(&T::lion, &T::tiger, &T::panda);
   };
};

struct zoo_t
{
   animals_t animals{};
   std::string name{"My Awesome Zoo"};

   struct glaze
   {
      using T = zoo_t;
      static constexpr auto value = glz::object(&T::animals, &T::name);
   };
};

"partial write"_test = [] {
      static constexpr auto partial = glz::json_ptrs("/name", "/animals/tiger");

      zoo_t obj{};
      std::string s{};
      const auto ec = glz::write_json<partial>(obj, s);
      expect(!ec);
      expect(s == R"({"animals":{"tiger":"Tiger"},"name":"My Awesome Zoo"})") << s;
   };
```
